htlcswitch/htlcnotifier/htlcnotifier.go

```diff
@@ -0,0 +1,262 @@
+// Package htlcnotifier notifies clients of HTLC forwards, failures and settles
+// for HTLCs that our node handles. HTLCs which originated at our node can be
+// identified by a zero channel ID, on the incoming channel for sends and on the
+// outgoing channel for receives.
+//
+// It takes subscriptions for its events and notifies them when HTLC events
+// occur. These are served on a best-effort basis; events are not persisted,
+// delivery is not guaranteed (in the event of a crash in the switch, forward
+// events may be lost) and events will be replayed upon restart. Events consumed
+// from this package should be de-duplicated, and not relied upon for critical
+// operations.
```

Comment:
From a greater perspective, I think it is still unsatisfactory that we are not building a reliable log of all decisions that we made. We are saving huge amounts of historical channel data in the database (including the raw onion payloads), but it is just not enough and/or not in the right format to extract what we want from it.

I believe that eventually there will be no way around creating a airtight audit trail.

Reply:
I don't think we are tying ourselves to anything that can't be fleshed out with a more reliable audit trail at a later stage?

I think that this best effort approach will provide node operators with value add, specifically if we add it to lndmon and get it as visible as possible. This is also a good opportunity for us to get some user feedback on the feature (specifically forward failures). Maybe people only care about actionable failures like `InsufficientBalance` (because they can potentially rebalance/loop), maybe they don't care about local events, hard to say till we get something usable out.

Reply:
Yes, I see that. I guess I am missing a bit the grand master plan of where we eventually want all of this to end up, but this PR could also be a good way to get a feel for requirements in that area.

Reply:
I think this PR specifically is about exposing useful information to routing operators. And there could very well be no master plan specifically in this region. We expose this information, put it in lndMon so it's nice and visible and people can start to understand their routing activities more. Best effort (which is actually pretty good, barring a crash in the switch) is sufficient. 

To me, this PR is about surfacing failures, because the rest of the information is already available (in forwarder log and list invoices/payments). In terms of an audit log, I do not think we are at the stage where we need to worry about auditing HTLC fails (yet). We need to start with the basics, auditing channel balances against the successful events in the channel and hardening the forwarder log.